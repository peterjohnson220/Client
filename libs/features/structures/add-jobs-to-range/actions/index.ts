export * from './jobs-to-grade.actions';
