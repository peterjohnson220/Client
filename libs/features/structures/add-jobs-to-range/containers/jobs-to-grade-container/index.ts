export * from './jobs-to-grade-container.component';
