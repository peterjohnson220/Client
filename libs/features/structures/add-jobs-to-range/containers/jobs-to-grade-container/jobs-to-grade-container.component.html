
<div #jobsToGrade class="jobs-to-grade-container">
  <table class="table default k-widget k-grid">
    <thead class="k-header pf-data-grid-header text-center">
    <th class="text-left">Grades</th>
    </thead>
  </table>
  <pf-async-container
    [loading]="loadingGrades$ | async"
    [loadingError]="error$ | async"
    [spinnerType]="spinnerType">

    <ng-container *ngIf="(grades$ | async)?.length; else noGradesMessage">
      <pf-job-to-grade
        *ngFor="let grade of grades$ | async; trackBy: trackByGradeId"
        [grade]="grade"
        [dragging]="isDragging$ | async"
        [rate]="rate"
        [rangeDistributionType]="gradeRangeGroupDetails.RangeDistributionTypeId"
        (loadJobs)="handleLoadJobs($event)"
        (jobDeleted)="handleJobDeleted($event)">
      </pf-job-to-grade>
    </ng-container>
  </pf-async-container>
</div>

<ng-template #noGradesMessage>
  <div *ngIf="!(loadingGrades$ | async)" class="no-results-message-container">
    No valid jobs selected. Please try again.
  </div>
</ng-template>
