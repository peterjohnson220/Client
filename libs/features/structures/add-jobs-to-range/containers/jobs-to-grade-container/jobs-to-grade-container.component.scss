@import 'assets/scss/legacy/legacy-variables';
@import '../../../../search/search/styles/shared-variables';

:host {
  @extend %position-below-results-header;
  width: 50%;
  right: 0;
  bottom: 15px;
  height: 90%;
}

.jobs-to-grade-container{
  position: absolute;
  margin-top: .5rem;
  margin-bottom: .5rem;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  overflow-y: scroll;
}

.no-results-message-container{
  display: flex;
  justify-content: center;
  margin-top: 10px;
  font-weight: $legacy-bold-font-weight;
}

.loading-more-row{
  height: 50px;
}

