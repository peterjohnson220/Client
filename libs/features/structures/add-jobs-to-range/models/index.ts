export * from './grade.model';
export * from './grade-job.model';
export * from './grade-range-group-details.model';
export * from './get-add-grade-jobs.model';
export * from './grade-jobs.model';
