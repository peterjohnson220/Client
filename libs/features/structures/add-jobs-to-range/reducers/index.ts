import { createSelector, createFeatureSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromJobsToGradeReducer from './jobs-to-grade.reducer';

import { AddJobsState } from '../../../jobs/add-jobs/reducers';

// Feature area state
export interface AddJobsToRangeState extends AddJobsState {
  jobsToGrade: fromJobsToGradeReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  feature_addJobsToRange: AddJobsToRangeState;
}

// Feature area reducers
export const reducers = {
  jobsToGrade: fromJobsToGradeReducer.reducer
};

// Select Feature Area
export const selectFeatureAreaState = createFeatureSelector<AddJobsToRangeState>('feature_addJobsToRange');

// Feature Selectors
export const selectJobsToGradeState = createSelector(
  selectFeatureAreaState,
  (state: AddJobsToRangeState) => state.jobsToGrade
);

// Jobs to Grade Page Selectors


export const getLoadingGrades = createSelector(
  selectJobsToGradeState,
  fromJobsToGradeReducer.getLoadingGrades
);

export const getGrades = createSelector(
  selectJobsToGradeState,
  fromJobsToGradeReducer.getGrades
);

export const getLoadingGradesError = createSelector(
  selectJobsToGradeState,
  fromJobsToGradeReducer.getLoadingGradesError
);



