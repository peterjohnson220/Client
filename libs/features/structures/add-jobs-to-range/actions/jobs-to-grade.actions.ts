import { Action } from '@ngrx/store';

import { Grade, GradeRangeGroupDetails, GradeJobs, GetGradeJobsModel, GradeJob } from '../models';

export const GET_GRADES = '[Project Add Jobs to Grade] Get Grades';
export const GET_GRADES_SUCCESS = '[Project Add Jobs to Grade] Get Grades Success';
export const GET_GRADES_ERROR = '[Project Add Jobs to Grade] Get Grades Error';
export const GET_GRADE_JOBS = '[Project Add Jobs to Grade] Get Grade Jobs';
export const GET_GRADE_JOBS_SUCCESS = '[Project Add Jobs to Grade] Get Grade Jobs Success';
export const GET_GRADE_JOBS_ERROR = '[Project Add Jobs to Grade] Get Grade Jobs Error';
export const ADD_JOBS_TO_GRADE = '[Project Add Jobs to Grade] Add New Jobs To Grade';
export const REMOVE_JOB = '[Project Add Jobs to Grade] Remove Job From Grade';
export const SAVE_GRADE_JOB_MAPS = '[Project Add Jobs to Grade] Save Grade Job Maps';
export const SAVE_GRADE_JOB_MAPS_SUCCESS = '[Project Add Jobs to Grade] Save Grade Job Maps Success';
export const SAVE_GRADE_JOB_MAPS_ERROR = '[Project Add Jobs to Grade] Save Grade Job Maps Error';
export const AUTO_GRADE_JOBS = '[Project Add Jobs to Grade] Auto Grade Jobs';
export const AUTO_GRADE_JOBS_SUCCESS = '[Project Add Jobs to Grade] Auto Grade Jobs Success';
export const AUTO_GRADE_JOBS_ERROR = '[Project Add Jobs to Grade] Auto Grade Jobs Error';

export class GetGrades implements Action {
  readonly type = GET_GRADES;

  constructor(public payload: GradeRangeGroupDetails) {}
}

export class GetGradesSuccess implements Action {
  readonly type = GET_GRADES_SUCCESS;

  constructor(public payload: Grade[]) {}
}

export class GetGradesError implements Action {
  readonly type = GET_GRADES_ERROR;

  constructor() {}
}

export class GetGradeJobs implements Action {
  readonly type = GET_GRADE_JOBS;

  constructor(public payload: GetGradeJobsModel) {}
}

export class GetGradeJobsError implements Action {
  readonly type = GET_GRADE_JOBS_ERROR;

  constructor(public payload: GradeJobs) {}
}

export class GetGradeJobsSuccess implements Action {
  readonly type = GET_GRADE_JOBS_SUCCESS;

  constructor(public payload: GradeJobs) {}
}

export class AddJobsToGrade implements Action {
  readonly type = ADD_JOBS_TO_GRADE;

  constructor(public payload: GradeJobs) {}
}

export class RemoveJob implements Action {
  readonly type = REMOVE_JOB;

  constructor(public payload: {GradeId: number, Job: GradeJob}) {}
}

export class SaveGradeJobMaps implements Action {
  readonly type = SAVE_GRADE_JOB_MAPS;

  constructor(public payload: Grade[]) {}
}

export class SaveGradeJobMapsSuccess implements Action {
  readonly type = SAVE_GRADE_JOB_MAPS_SUCCESS;

  constructor() {}
}

export class SaveGradeJobMapsError implements Action {
  readonly type = SAVE_GRADE_JOB_MAPS_ERROR;

  constructor() {}
}

export class AutoGradeJobs implements Action {
  readonly type = AUTO_GRADE_JOBS;

  constructor(public gradeRangeGroupDetails: GradeRangeGroupDetails) {}
}

export class AutoGradeJobsSuccess implements Action {
  readonly type = AUTO_GRADE_JOBS_SUCCESS;

  constructor() {}
}

export class AutoGradeJobsError implements Action {
  readonly type = AUTO_GRADE_JOBS_ERROR;

  constructor() {}
}

export type JobsToGradeActions
  = GetGrades
  | GetGradesSuccess
  | GetGradesError
  | GetGradeJobs
  | GetGradeJobsSuccess
  | GetGradeJobsError
  | AddJobsToGrade
  | RemoveJob
  | SaveGradeJobMaps
  | SaveGradeJobMapsSuccess
  | SaveGradeJobMapsError
  | AutoGradeJobs
  | AutoGradeJobsSuccess
  | AutoGradeJobsError;
