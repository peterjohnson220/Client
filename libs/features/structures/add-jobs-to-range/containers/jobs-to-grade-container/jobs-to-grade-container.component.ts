import { Component, OnDestroy, ViewChildren, QueryList, Input, OnInit } from '@angular/core';

import { Store } from '@ngrx/store';
import { Observable, Subscription, merge } from 'rxjs';
import { mapTo, startWith } from 'rxjs/operators';
import { DragulaService } from 'ng2-dragula';
import * as autoScroll from 'dom-autoscroller';

import * as fromAddJobsReducer from 'libs/features/jobs/add-jobs/reducers';
import * as fromAddJobsActions from 'libs/features/jobs/add-jobs/actions/search-results.actions';

import { JobToGradeComponent } from '../../components';
import * as fromJobsToGradeReducer from '../../reducers';
import * as fromJobsToGradeActions from '../../actions';
import { Grade, GradeRangeGroupDetails } from '../../models';

@Component({
  selector: 'pf-jobs-to-grade-container',
  templateUrl: './jobs-to-grade-container.component.html',
  styleUrls: ['./jobs-to-grade-container.component.scss']
})
export class JobsToGradeContainerComponent implements OnDestroy, OnInit {
  @ViewChildren(JobToGradeComponent) jobsToGradeComponents !: QueryList<JobToGradeComponent>;
  @Input() gradeRangeGroupDetails: GradeRangeGroupDetails;
  @Input() rate: string;
  @Input() controlPoint: string;
  // Observables
  grades$: Observable<Grade[]>;
  selectedJobs$: Observable<string[]>;
  loadingGrades$: Observable<boolean>;
  error$: Observable<boolean>;
  isDragging$: Observable<boolean>;
  dragStart$: Observable<boolean>;
  dragEnd$: Observable<boolean>;

  // Subscriptions
  dragSubs: Subscription;
  selectedJobsSubscription: Subscription;

  selectedJobs: number[];
  scroll: any;
  isDragging: boolean;

  spinnerType = 'GIF';

  constructor(
    private store: Store<fromJobsToGradeReducer.State>,
    private dragulaService: DragulaService
  ) {
    this.selectedJobs = [];
    this.dragSubs = new Subscription();
    this.selectedJobsSubscription = new Subscription();
    this.selectedJobs$ = this.store.select(fromAddJobsReducer.getSelectedJobIds);
    this.grades$ = this.store.select(fromJobsToGradeReducer.getGrades);
    this.loadingGrades$ = this.store.select(fromJobsToGradeReducer.getLoadingGrades);
    this.error$ = this.store.select(fromJobsToGradeReducer.getLoadingGradesError);

    this.selectedJobsSubscription = this.selectedJobs$.subscribe(jobs => {
     this.selectedJobs = jobs.map(Number);
      // TODO: come up with a better solution to initialize the auto scroll.
      // These elements are not on the DOM until the multi match modal opens in Client implementations
     if (!this.scroll && jobs.length) {
       this.initializeScroll();
     }
    });
    this.configureDragEvents();
  }

  trackByGradeId(index, item: Grade) {
    return item.CompanyStructuresGradesId;
  }

  handleLoadJobs(grade: Grade): void {
    this.store.dispatch(new fromJobsToGradeActions.GetGradeJobs(
      { CompanyStructuresGradesId: grade.CompanyStructuresGradesId, ControlPoint: this.controlPoint,
        CompanyStructuresRangeGroupId: grade.CompanyStructuresRangeGroupId, JobIds: [] }));
  }

  private initializeScroll() {
    const that = this;
    this.scroll = autoScroll(
      document.querySelector('.jobs-to-grade-container'),
      {
        margin: 30,
        maxSpeed: 25,
        scrollWhenOutside: true,
        autoScroll: function() {
          return this.down && that.isDragging;
        }
      });
  }

  private configureDragEvents(): void {
    this.dragStart$ = this.dragulaService.drag('jobs-grade-bag').pipe(mapTo(true));
    this.dragEnd$ = this.dragulaService.dragend('jobs-grade-bag').pipe(mapTo(false));
    this.isDragging$ = merge(this.dragStart$, this.dragEnd$).pipe(startWith(false));
    this.dragSubs.add(this.dragulaService.over('jobs-grade-bag')
      .subscribe(({ el, container }) => {
        if (container.classList.contains('dropzone')) {
          container.classList.add('highlight');
        }
      })
    );
    this.dragSubs.add(this.dragulaService.out('jobs-grade-bag')
      .subscribe(({ el, container }) => {
        if (container.classList.contains('dropzone')) {
          container.classList.remove('highlight');
        }
      })
    );
    this.dragSubs.add(this.dragulaService.drop('jobs-grade-bag')
        .subscribe(({ el, target }) => {
          target.classList.remove('highlight');
          const gradeIdAttribute = target.attributes['data-grade-id'];
          if (gradeIdAttribute && gradeIdAttribute.value) {
            const gradeId = Number(gradeIdAttribute.value);
            this.store.dispatch(new fromJobsToGradeActions.AddJobsToGrade(
              { CompanyStructuresRangeGroupId: this.gradeRangeGroupDetails.RangeGroupId,
                CompanyStructuresGradesId: gradeId, JobIds: this.selectedJobs, ControlPoint: this.controlPoint }));
            this.store.dispatch(new fromAddJobsActions.ClearSelectedJobs());
            this.showJobsOnGrade(gradeId);
          }
        el.innerHTML = '';
        })
    );

    this.isDragging$.subscribe(dragStatus => {
      this.isDragging = dragStatus;
    });
  }

  ngOnDestroy(): void {
    this.dragSubs.unsubscribe();
    this.selectedJobsSubscription.unsubscribe();
  }

  ngOnInit() {
    this.store.dispatch(new fromJobsToGradeActions.GetGrades(this.gradeRangeGroupDetails));
  }

  showJobsOnGrade(gradeId: number) {
    const gradeComponents = this.jobsToGradeComponents.toArray();
    const matchingComponent = gradeComponents.find(c => c.grade.CompanyStructuresGradesId === gradeId);
    if (matchingComponent) {
      matchingComponent.showJobsDisplay();
    }
  }
}

