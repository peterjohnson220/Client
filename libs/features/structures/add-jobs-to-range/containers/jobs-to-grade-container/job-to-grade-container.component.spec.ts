import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { combineReducers, Store, StoreModule } from '@ngrx/store';
import { DragulaModule } from 'ng2-dragula';
import { of } from 'rxjs/index';

import * as fromRootState from 'libs/state/state';

import { JobsToGradeContainerComponent } from './jobs-to-grade-container.component';
import { generateMockGrade, generateMockGetGradeJobsModel, generateMockGradeJob } from '../../models';
import * as fromAddJobsReducer from '../../reducers';
import * as fromJobsToGradeActions from '../../actions/jobs-to-grade.actions';

describe('Project - AddJobs - JobToGrade Container Component', () => {
  let instance: JobsToGradeContainerComponent;
  let fixture: ComponentFixture<JobsToGradeContainerComponent>;
  let store: Store<fromAddJobsReducer.State>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          feature_addJobsToRange: combineReducers(fromAddJobsReducer.reducers),

        }),
        DragulaModule.forRoot()
      ],
      declarations: [ JobsToGradeContainerComponent ],
      schemas: [ NO_ERRORS_SCHEMA ]
    });
  });

  beforeEach(() => {

    store = TestBed.inject(Store);
    fixture = TestBed.createComponent(JobsToGradeContainerComponent);
    instance = fixture.componentInstance;
    instance.controlPoint = 'BaseMRP';

  });

  it('should get jobs when load jobs clicked', () => {
    const model = generateMockGetGradeJobsModel();
    const expectedAction = new fromJobsToGradeActions.GetGradeJobs(model);
    spyOn(store, 'dispatch');

    instance.handleLoadJobs(generateMockGrade());
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();

    expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
  });

  it('should display grades instead of the no valid jobs message when grades are found', () => {
    instance.grades$ = of([generateMockGrade()]);

    expect(fixture).toMatchSnapshot();


  });

  it('should not get jobs when load jobs clicked but jobs are already loaded', () => {
    const model = generateMockGetGradeJobsModel();
    const expectedAction = new fromJobsToGradeActions.GetGradeJobs(model);
    spyOn(store, 'dispatch');

    const grade = generateMockGrade();
    grade.Jobs = [generateMockGradeJob()];
    instance.handleLoadJobs(grade);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();

    expect(store.dispatch).not.toHaveBeenCalledWith(expectedAction);
  });

  it('should dispatch a delete job action when job deleted', () => {
    const grade = generateMockGrade();
    grade.Jobs = [generateMockGradeJob()];
    const expectedAction = new fromJobsToGradeActions.RemoveJob({GradeId: grade.CompanyStructuresGradesId, Job: grade.Jobs[0]});
    spyOn(store, 'dispatch');

    instance.handleJobDeleted({ job: grade.Jobs[0], grade: grade});
    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
  });

});
