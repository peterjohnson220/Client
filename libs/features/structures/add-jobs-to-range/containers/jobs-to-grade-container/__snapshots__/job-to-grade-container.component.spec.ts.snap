// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project - AddJobs - JobToGrade Container Component should display grades instead of the no valid jobs message when grades are found 1`] = `
<pf-jobs-to-grade-container
  controlPoint={[Function String]}
  dragEnd$={[Function AnonymousSubject]}
  dragStart$={[Function AnonymousSubject]}
  dragSubs={[Function Subscription]}
  dragulaService={[Function DragulaService]}
  error$={[Function Store]}
  grades$={[Function Observable]}
  isDragging="false"
  isDragging$={[Function Observable]}
  loadingGrades$={[Function Store]}
  selectedJobs={[Function Array]}
  selectedJobs$={[Function Store]}
  selectedJobsSubscription={[Function Subscriber]}
  spinnerType={[Function String]}
  store={[Function Store]}
>
  <div
    class="jobs-to-grade-container"
  >
    <table
      class="table default k-widget k-grid"
    >
      <thead
        class="k-header pf-data-grid-header text-center"
      >
        <th
          class="text-left"
        >
          Grades
        </th>
      </thead>
    </table>
    <pf-async-container>
      
    </pf-async-container>
  </div>
</pf-jobs-to-grade-container>
`;

exports[`Project - AddJobs - JobToGrade Container Component should get jobs when load jobs clicked 1`] = `
<pf-jobs-to-grade-container
  controlPoint={[Function String]}
  dragEnd$={[Function AnonymousSubject]}
  dragStart$={[Function AnonymousSubject]}
  dragSubs={[Function Subscription]}
  dragulaService={[Function DragulaService]}
  error$={[Function Store]}
  grades$={[Function Store]}
  isDragging="false"
  isDragging$={[Function Observable]}
  jobsToGradeComponents={[Function QueryList]}
  loadingGrades$={[Function Store]}
  selectedJobs={[Function Array]}
  selectedJobs$={[Function Store]}
  selectedJobsSubscription={[Function Subscriber]}
  spinnerType={[Function String]}
  store={[Function Store]}
>
  <div
    class="jobs-to-grade-container"
  >
    <table
      class="table default k-widget k-grid"
    >
      <thead
        class="k-header pf-data-grid-header text-center"
      >
        <th
          class="text-left"
        >
          Grades
        </th>
      </thead>
    </table>
    <pf-async-container>
      
      
      <div
        class="no-results-message-container"
      >
         No valid jobs selected. Please try again. 
      </div>
      
    </pf-async-container>
  </div>
</pf-jobs-to-grade-container>
`;

exports[`Project - AddJobs - JobToGrade Container Component should not get jobs when load jobs clicked but jobs are already loaded 1`] = `
<pf-jobs-to-grade-container
  controlPoint={[Function String]}
  dragEnd$={[Function AnonymousSubject]}
  dragStart$={[Function AnonymousSubject]}
  dragSubs={[Function Subscription]}
  dragulaService={[Function DragulaService]}
  error$={[Function Store]}
  grades$={[Function Store]}
  isDragging="false"
  isDragging$={[Function Observable]}
  jobsToGradeComponents={[Function QueryList]}
  loadingGrades$={[Function Store]}
  selectedJobs={[Function Array]}
  selectedJobs$={[Function Store]}
  selectedJobsSubscription={[Function Subscriber]}
  spinnerType={[Function String]}
  store={[Function Store]}
>
  <div
    class="jobs-to-grade-container"
  >
    <table
      class="table default k-widget k-grid"
    >
      <thead
        class="k-header pf-data-grid-header text-center"
      >
        <th
          class="text-left"
        >
          Grades
        </th>
      </thead>
    </table>
    <pf-async-container>
      
      
      <div
        class="no-results-message-container"
      >
         No valid jobs selected. Please try again. 
      </div>
      
    </pf-async-container>
  </div>
</pf-jobs-to-grade-container>
`;
